src/Raven.Server/Documents/Indexes/Analysis/AnalyzerCompilationCache.cs

```diff
@@ -0,0 +1,158 @@
+﻿using System;
+using System.Collections.Concurrent;
+using Raven.Client.Documents.Indexes.Analysis;
+using Raven.Client.ServerWide;
+using Raven.Server.Utils;
+
+namespace Raven.Server.Documents.Indexes.Analysis
+{
+    public static class AnalyzerCompilationCache
+    {
+        private static readonly ConcurrentDictionary<CacheKey, Lazy<Type>> AnalyzersCache = new ConcurrentDictionary<CacheKey, Lazy<Type>>();
+
+        private static readonly ConcurrentDictionary<CacheKey, Lazy<Type>> AnalyzersPerDatabaseCache = new ConcurrentDictionary<CacheKey, Lazy<Type>>();
+
+        public static Type GetAnalyzerType(string name, string databaseName)
```

Comment:
I think it would be better to use the hash of the analyzer text here, instead.
These are going to live forever, and if we have multiple databases, an analzyer is going to be used across many of them.

For that matter, can I define a default custom analyzer?

Reply:
That would also save the compilation costs, the need to remove analyzers, etc.

Reply:
There are two caches here. One global-level to avoid compiling same thing all over. Second per database so users cannot use analyzers that were not meant to be used in this database. Identical mechanism for custom sorters.

Yes, you can use configuration setting to define a default analyzer for a given database/server. Now when I'm thinking about that, you cannot define sorters/analyzers server-wide. You are suggesting to add this ability?

Reply:
The issue is that we now have a static cache that never expires.
When are you removing from `AnalyzersPerDatabaseCache ` ?

Reply:
This is handled in `public static void AddAnalyzers(DatabaseRecord databaseRecord)` we are removing analyzers that are no longer in the DatabaseRecord there.

Reply:
And when the database is removed?